bazel/bazel.go

```diff
@@ -79,6 +79,13 @@ func bazeliskNpmPath(ibazelBinPath string) (string, error) {
 				ext = ".exe"
 			}
 			name := strings.Join(append(prefix, nm, scope, "bazelisk", "bazelisk-"+bin+ext), "/")
+			_, err := os.Stat(name)
+			if err != nil {
+				if !os.IsNotExist(err) {
+					return "", err
+				}
+				continue
```

Comment:
This entire loop seems strange. I it looks like it is taking /a/long/path/like/this/ending/with/node_modules/@bazel/... and we are searching pulling out 5 parts at a time just to check.
Is that so? Why aren't we just scanning for each occurance of  node_modules before assigning the intermedates.

It feels like this  should not be a loop. Simply split the string and look at the last 6 components.
github, alas, does not let me add the comment at line 74, where it belongs.

Reply:
Hey @aiuto, thanks for the review. That's a great question. I think it's because bazelisk might be installed as a peer dependency, or as a dependency at a higher level. This code was copied from the `bazelNpmPath`, but I should've factored it out then.

I think it's because, they can be peer dependencies, like:

```
// /DIR/node_modules/@bazel/ibazel/bin/darwin_amd64/ibazel
// /DIR/node_modules/@bazel/bazel-darwin_x64/bazel-0.28.0-darwin-x86_64
```

or ibazel could be a deeper dependency

```
// /DIR/node_modules/DIR_2/node_modules/@bazel/ibazel/bin/darwin_amd64/ibazel
// /DIR/node_modules/@bazel/bazel-darwin_x64/bazel-0.28.0-darwin-x86_64
```

Reply:
This isn't really a hot loop, so I don't think the efficiency matters too much, but I'll take a look at doing it with that logic.

Reply:
@aiuto to confirm what @zoidbergwill is flicking at, the recursive nature of NPM's distribution make things... hard to say the least. This code is only invoked once per run and it is worth spending the cycles looping through the filesystem to ensure correct behavior above speed. iBazel is not augmented with telemetry to report on user behavior. Assuming my use is representative (and I know it isn't), I launch iBazel a couple of times over the course of an hour and it remains running for something on the order of 15 minutes with this loop representing an infinitesimally small fraction of that time.

Maybe the action here is to learn from the design decisions of others and not adopt this as a strategy for the core of Bazel. NPM is just too complex.

I presume you spend a great deal of time interacting bazel as a part of your job. Maybe ibazel could be added as a tool in your toolkit. It's not too fancy. It simply rdeps queries the target you want to `build`, `run`, or `test, and watches the inodes of any inputs. When those inodes broadcast that they have changed it reruns `bazel` with the action you have requested on the targets you provided. I'd love to have you try it out and tell me some ways I can improve it. What do you think about that? 

@zoidbergwill, I want to again praise you for contributing to open source software. Thank you for volunteering your time and investing your efforts. For me, knowing that people are using something I built, especially when it's in a way that I didn't anticipate like the TS/JS community is doing, brings me joy. Thank you so much!

Reply:
So I tried to simplify it with logic that looked something like the following, but I don't know how much it helps with the readability:

```
+// findAllNodeModules helps us find where bazel/bazelisk might be installed
+// if ibazel lives here:
+// /DIR/node_modules/package_without_namespace/node_modules/@namespace/package/node_modules/@bazel/ibazel/bin/darwin_amd64/ibazel
+// We could find bazel in
+// /DIR/node_modules/@bazel/bazel-darwin_x64/bazel-0.28.0-darwin-x86_64
+// /DIR/node_modules/package_without_namespace/node_modules/@bazel/bazel-darwin_x64/bazel-0.28.0-darwin-x86_64
+// /DIR/node_modules/package_without_namespace/node_modules/@namespace/package/node_modules/@bazel/bazel-darwin_x64/bazel-0.28.0-darwin-x86_64
+func findAllNodeModules(binPath string) []string {
+       var results []string
+       s := strings.Split(binPath, string(os.PathSeparator))
+       for i, c := range s {
+               if c == "node_modules" {
+                       results = append(results, filepath.Join(s[:i]...))
+               }
+       }
+       return results
+}
+
```

Reply:
I'm fine with either. Again, the lack of an example string in the comment (showing that node_modules can appear many times) threw me off.